Automattic/jetpack
# [CRM] Add "Create Workflow" endpoint

Add an endpoint that can create automation workflows, one at a time.

Fixes https://github.com/Automattic/zero-bs-crm/issues/3348

## Proposed changes:

* Add `[POST] /wp-json/jetpack-crm/v4/automation/workflows` endpoint.
* Add test to verify that the endpoint works.

### Other information:


## Jetpack product discussion



* pbhBOz-3ss-p2
* pbhBOz-2ES-p2
* pbhBOz-3uX-p2

## Does this pull request change what data or activity we track or use?



No.

## Testing instructions:






### Automated tests

All of the tests should obviously pass, but looking at the following automated tests should suffice:

* `REST_Authentication_Test`: Updated to include our new endpoint.
* `REST_Automation_Workflows_Controller_Test`: Updated with a new test to verify the endpoint works.

### Manual testing

**Manual testing prep work**

1. Make sure the following feature flags are enabled on your site:

```php
add_filter( 'jetpack_crm_feature_flag_api_v4', '__return_true' );
add_filter( 'jetpack_crm_feature_flag_automations', '__return_true' );
```

2. Modify `REST_Automation_Workflows_Controller::get_item_permissions_check` to return `true` at the top.

_We do this to make it easier to call the endpoints. This should allow you to do it from any client you want (e.g. Postman) without having to worry about WP Logged in cookies, nonce, etc._

**Create Workflow**

* Method: `POST`
* Path: `/wp-json/jetpack-crm/v4/automation/workflows`
* Payload: `{"name":"Workflow Test: with_condition_action","description":"Test: the description of the workflow","category":"Test","active":true,"triggers":["jpcrm\/contact_created"],"initial_step":0,"steps":[{"slug":"contact_status_condition","class_name":"Automattic\\Jetpack\\CRM\\Automation\\Tests\\Mocks\\Contact_Condition","attributes":{"field":"status","operator":"is","value":"lead"},"next_step_true":1,"next_step_false":null}]}`
